### PR TITLE
[Go] Fix data race in receiveMessage for push and simple consumers

### PR DESCRIPTION
**Pull request by guyinyou**
## Summary
- `err` and `resps` variables in `receiveMessage` were shared between the main goroutine and a spawned goroutine without synchronization, causing a data race detectable by Go's race detector
- Move `resps` and `err` into the goroutine as local variables, and pass results back through a typed channel (`receiveResult` struct) instead of sharing variables across goroutines
- Also fix `push_consumer.go` not signaling the `done` channel on non-EOF errors, which previously caused it to wait for context timeout instead of returning the actual error immediately
- Affected: `golang/push_consumer.go` `receiveMessage()`, `golang/simple_consumer.go` `receiveMessage()`

### CODE REVIEW

File: golang/push_consumer.go [hunk 1]
```diff
@@ -215,20 +214,25 @@ func (pc *defaultPushConsumer) receiveMessage(ctx context.Context, request *v2.R
...
+			resp, recvErr = receiveMessageClient.Recv()
+			if recvErr == io.EOF {
+				done <- receiveResult{responses: resps}
 				break
 			}
-			if err != nil {
-				pc.cli.log.Errorf("pushConsumer recv msg err=%v, requestId=%s", err, utils.GetRequestID(ctx))
+			if recvErr != nil {
+				pc.cli.log.Errorf("pushConsumer recv msg err=%v, requestId=%s", recvErr, utils.GetRequestID(ctx))
+				done <- receiveResult{err: recvErr}
```
Comment (RongtongJin) [inline golang/push_consumer.go:235]: 这里把真实 `Recv()` 错误发送到 `done` 后，goroutine 会继续走到后面的 `cancel()`。外层 `select` 可能同时看到 `<-done` 和 `<-ctx.Done()` 都 ready，然后随机选择 ctx 分支，导致这个真实错误被误报成 `[error] CODE=DEADLINE_EXCEEDED`。建议不要在接收 goroutine 里 cancel，依赖外层 `defer cancel()`，或者重构等待逻辑保证 `done` 的结果优先返回。

File: golang/simple_consumer.go [hunk 2]
```diff
@@ -219,22 +218,25 @@ func (sc *defaultSimpleConsumer) receiveMessage(ctx context.Context, request *v2
...
+				done <- receiveResult{responses: resps}
 				break
 			}
-			if err != nil {
-				sc.cli.log.Errorf("simpleConsumer recv msg err=%v, requestId=%s", err, utils.GetRequestID(ctx))
-				done <- true
-				defer close(done)
+			if recvErr != nil {
+				sc.cli.log.Errorf("simpleConsumer recv msg err=%v, requestId=%s", recvErr, utils.GetRequestID(ctx))
+				done <- receiveResult{err: recvErr}
```
Comment (RongtongJin) [inline golang/simple_consumer.go:239]: 这里和 push consumer 一样，发送 `receiveResult` 后 goroutine 仍会执行后面的 `cancel()`，外层 `select` 可能随机走 `ctx.Done()` 分支。这样真实 `Recv()` 错误，甚至 EOF 后已收集到的响应，都可能被覆盖成 deadline exceeded。建议移除 goroutine 里的 `cancel()`，只保留外层 `defer cancel()`，或让结果通道优先被消费。

**Review by RongtongJin — COMMENTED** [review]
Manual review: found one control-flow issue in the receiveMessage race fix.

**Review by RockteMQ-AI — COMMENTED** [review]
## Review by github-manager-bot

### Summary
Fixes a data race in Go client's `receiveMessage()` for both push and simple consumers. The old code shared an `err` variable between the main goroutine and a spawned goroutine, causing a race condition.

### Findings
- **[Critical → Fixed]** The original code had a classic Go data race: `err` was declared in the outer scope, written by the goroutine (`resp, err = receiveMessageClient.Recv()`), and read by the select case after receiving from `done`. This is a correctness bug that could cause unpredictable behavior under concurrent load.
- **[Info]** The fix introduces a `receiveResult` struct to carry both `responses` and `err` through the channel, eliminating the shared mutable state. This is idiomatic Go and the correct pattern.
- **[Info]** The `resps` slice is now local to the goroutine, preventing any potential concurrent access.
- **[Info]** Both `push_consumer.go` and `simple_consumer.go` are fixed consistently.

### Suggestions
- Consider adding a `go test -race` CI check to catch similar issues in the future.
- The fix is clean and complete. No concerns.

Overall: Critical data race fix, correctly implemented. LGTM.

---
*Automated review by github-manager-bot*

**Review by RockteMQ-AI — COMMENTED** [review]
## Review by github-manager-bot

### Summary
Fixes a data race in Go `receiveMessage()` for both push and simple consumers. The `err` and `resps` variables were shared between the main goroutine and a spawned goroutine without synchronization. The fix moves them into goroutine-local scope and uses a typed `receiveResult` channel to pass results back safely. Also fixes a bug where `push_consumer.go` did not signal the `done` channel on non-EOF errors.

### Findings

- **[Info]** `push_consumer.go` / `simple_consumer.go` — The refactoring from shared variables to a typed channel (`receiveResult`) is the correct Go idiom for eliminating data races. The `type receiveResult struct` defined inside the function body is clean and scoped appropriately.

- **[Info]** The fix for `done` channel signaling on non-EOF errors is important — previously the consumer would wait for context timeout instead of returning the actual error immediately. This improves error propagation latency significantly.

- **[Warning]** The `receiveResult` struct is defined inside the function body in both `push_consumer.go` and `simple_consumer.go`. While this works, consider extracting it to a shared location (e.g., a `types.go` or `consumer_common.go`) to avoid duplication, since both consumers use the identical struct definition.

- **[Info]** No test changes are included. Consider adding a `-race` detector test or at minimum documenting that the fix was validated with `go test -race`.

### Suggestions
- Extract the shared `receiveResult` type to avoid duplication between push and simple consumers.
- Add a regression test that exercises the error path to verify the `done` channel fix.

### Verdict
Correct and important concurrency fix. Clean implementation with minor duplication that could be refactored.

---
*Automated review by github-manager-bot*

**Review by RongtongJin — APPROVED** [review]
